### DISCUSSION TITLE
Hook/event to update other record

### DISCUSSION THREAD

**Discussion by zgid123** [mikro-orm/mikro-orm#3648] [Q&A]
Hi,

Assume I have a model `Foo` has manyToMany association with `Bar` via `FoosBar` and has oneToMany with `C`.

When I create/update record for `C`, I wanna update one or more columns for `FoosBar` based on `Foo` id from `C`.

So, is there any hook or event can do it at the moment? For hook `afterCreate` and `afterUpdate`, we cannot use flush, so we will ignore them. `onFlush` seems not to work too - or I do it all wrong so it cannot update inside `onFlush`?

**Comment by rcfox** [comment]
Are you remembering to update the unit of work? 

https://mikro-orm.io/docs/events#using-onflush-event

**Reply by zgid123** [reply]
no, the event will be frozen because I need to query the correct record from table C to update its value

**Reply by B4nan** [reply]
You wont get much help if you dont share any code. onFlush is the way to go, it works just fine if you use it right. Show what have you tried.

**Reply by zgid123** [reply]
Sorry, I forgot to give the example. Here the example:

```ts
public async onFlush({ uow, em }: FlushEventArgs): Promise<void> {
    const changeSets = uow.getChangeSets();
    const cs = changeSets.find(({ type, entity }) => {
      return (
        [
          ChangeSetType.CREATE,
          ChangeSetType.UPDATE,
          ChangeSetType.DELETE,
        ].includes(type) && entity instanceof FoosBar
      );
    });

    if (!cs) {
      return;
    }

    const cRepo =
      em.getRepository<C>('C');
    const { foo } = cs.entity as FoosBar;
    const c = await cRepo.findOne({
      foo,
    });
    cRepo.assign(c, {
      columnA: 'value',
    });
  }
```
 
The context is whenever I create `FoosBar` record, I need to update one or more columns of `C`, and C is got using `foo` from `FoosBar` record. When I debug, the event will be stuck at the line `findOne`. Because I couldn't find a correct event, so I just tried the `onFlush`

**Reply by zgid123** [reply]
I solved this problem by `populate` `C` when getting `Foo`, and then create `FoosBar` with `Foo` as association. Inside the `onFlush`, I will loop the list `C` to get the correct record, then update it like the document guides.

```ts
// service.ts
@Injectable()
class Service {
  constructor(
    @InjectRepository(Foo)
    private readonly fooRepo: EntityRepository<Foo>,
    @InjectRepository(Bar)
    private readonly barRepo: EntityRepository<Bar>,
    @InjectRepository(FoosBar)
    private readonly foosBarRepo: EntityRepository<FoosBar>,
  ) {}

  function async exec({ fooId, barId, ...rest }): Promise<FoosBar> {
    const foo = await this.fooRepo.findOne(
      {
        id: fooId,
      },
      {
        populate: ['cs'],
      },
    );

    const bar = await this.barRepo.findOne({ id: barId });

    const foosBar =  this.foosBarRepo.create({
      ...rest,
      foo,
      bar,
    });

    await this.foosBarRepo.persist(foosBar).flush();

    return foosBar;
  }
}

// event
public async onFlush({ uow, em }: FlushEventArgs): Promise<void> {
  const changeSets = uow.getChangeSets();
  const cs = changeSets.find(({ type, entity }) => {
    return (
      [
        ChangeSetType.CREATE,
        ChangeSetType.UPDATE,
        ChangeSetType.DELETE,
      ].includes(type) && entity instanceof FoosBar
    );
  });

  if (!cs) {
    return;
  }

  const { foo, bar, valueForColumnA } = cs.entity as FoosBar;
  const c = foo.cs.find((element) => element.barId === bar.id);
  
  if (c) {
    Object.assign(c, { columnA: valueForColumnA });
  }

  uow.computeChangeSet(c);
  uow.recomputeSingleChangeSet(cs.entity);
}
```

**Reply by zgid123** [reply]
actually I don't wanna populate `cs` at the moment, because I don't use it anywhere later. But it seems whenever I run findOne here, it will freeze the app, so if we have a solution to get the correct record from database directly, please let me know.

Also, the onFlush will run twice, so I trick it by set a key-value for the changeSet to skip the second run

**Comment by superiums** [comment]
i've tried to call findOne on beforeCreate and it success and not  freeze the app.
